handler.go

```diff
@@ -24,7 +46,43 @@ func (s *server) handleToken(w http.ResponseWriter, r *http.Request) {
 		})
 	}
 
-	body, err := s.getAccessTokenPayload(ctx, cookies)
+	key := cookiesKey(cookies)
+
+	// --- Cache check ---
+	s.cacheMutex.Lock()
+	entry, exists := s.tokenCache[key]
+	if exists && time.Now().Before(entry.Expiry) {
+		s.cacheMutex.Unlock()
+		slog.InfoContext(ctx, "Returning cached Spotify token for key", slog.String("key", key))
+		w.Header().Set("Content-Type", "application/json")
+		_, _ = w.Write(entry.Token)
+		return
+	}
+	s.cacheMutex.Unlock()
+	// -------------------
+
+	// Use singleflight to prevent duplicate fetches
+	v, err, _ := s.group.Do(key, func() (interface{}, error) {
```

Comment:
pretty sure you should call this before you do any processing for it to work correctly
we also don't use `interface{}` anymore